# 社会科学のためのデータ分析入門（上） chpater 1.5.1
source: http://qss.princeton.press/student-files/INTRO/turnout.csv

In [1]:
import pandas as pd

In [2]:
table = pd.read_csv('turnout.csv')

## 1. データを可視化する

In [3]:
table

year     VEP     VAP   total  ANES  felons  noncit  overseas  osvoters
0   1980  159635  164445   86515    71     802    5756      1803       NaN
1   1982  160467  166028   67616    60     960    6641      1982       NaN
2   1984  167702  173995   92653    74    1165    7482      2361       NaN
3   1986  170396  177922   64991    53    1367    8362      2216       NaN
4   1988  173579  181955   91595    70    1594    9280      2257       NaN
5   1990  176629  186159   67859    47    1901   10239      2659       NaN
6   1992  179656  190778  104405    75    2183   11447      2418       NaN
7   1994  182623  195258   75106    56    2441   12497      2229       NaN
8   1996  186347  200016   96263    73    2586   13601      2499       NaN
9   1998  190420  205313   72537    52    2920   14988      2937       NaN
10  2000  194331  210623  105375    73    3083   16218      2937       NaN
11  2002  198382  215462   78382    62    3168   17237      3308       NaN
12  2004  203483  220336  122295    77    3158   18068      3862       NaN
13  2008  213314  230872  131304    78    3145   19392      4972     263.0

VEP：有権者人口

VAP：選挙年齢人口（選挙権不保持の者も含む）

In [4]:
table.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   year      14 non-null     int64  
 1   VEP       14 non-null     int64  
 2   VAP       14 non-null     int64  
 3   total     14 non-null     int64  
 4   ANES      14 non-null     int64  
 5   felons    14 non-null     int64  
 6   noncit    14 non-null     int64  
 7   overseas  14 non-null     int64  
 8   osvoters  1 non-null      float64
dtypes: float64(1), int64(8)
memory usage: 1.1 KB


In [5]:
table.describe()

year            VEP            VAP          total       ANES  \
count    14.000000      14.000000      14.000000      14.000000  14.000000   
mean   1993.142857  182640.285714  194225.857143   89778.285714  65.785714   
std       8.618916   16065.143680   20543.936013   20627.893112  10.474721   
min    1980.000000  159635.000000  164445.000000   64991.000000  47.000000   
25%    1986.500000  171191.750000  178930.250000   73179.250000  57.000000   
50%    1993.000000  181139.500000  193018.000000   89055.000000  70.500000   
75%    1999.500000  193353.250000  209295.500000  102369.500000  73.750000   
max    2008.000000  213314.000000  230872.000000  131304.000000  78.000000   

            felons        noncit     overseas  osvoters  
count    14.000000     14.000000    14.000000       1.0  
mean   2176.642857  12229.142857  2745.714286     263.0  
std     875.275131   4470.780002   840.053975       NaN  
min     802.000000   5756.000000  1803.000000     263.0  
25%    1423.750000   8591.500000  2236.000000     263.0  
50%    2312.000000  11972.000000  2458.500000     263.0  
75%    3042.250000  15910.500000  2937.000000     263.0  
max    3168.000000  19392.000000  4972.000000     263.0

14個の観察があり、1980年から2008年までの投票率に関するデータが含まている。

## 2. 投票率を求める

In [6]:
table['VEPratio'] = table['total'] / table['VEP'] * 100
table['VAPratio'] = table['total'] / (table['VAP']+table['overseas']) * 100

In [7]:
table.head()

year     VEP     VAP  total  ANES  felons  noncit  overseas  osvoters  \
0  1980  159635  164445  86515    71     802    5756      1803       NaN   
1  1982  160467  166028  67616    60     960    6641      1982       NaN   
2  1984  167702  173995  92653    74    1165    7482      2361       NaN   
3  1986  170396  177922  64991    53    1367    8362      2216       NaN   
4  1988  173579  181955  91595    70    1594    9280      2257       NaN   

    VEPratio   VAPratio  
0  54.195509  52.039724  
1  42.137013  40.245223  
2  55.248596  52.537481  
3  38.141154  36.078451  
4  52.768480  49.722602

In [8]:
table.describe()

year            VEP            VAP          total       ANES  \
count    14.000000      14.000000      14.000000      14.000000  14.000000   
mean   1993.142857  182640.285714  194225.857143   89778.285714  65.785714   
std       8.618916   16065.143680   20543.936013   20627.893112  10.474721   
min    1980.000000  159635.000000  164445.000000   64991.000000  47.000000   
25%    1986.500000  171191.750000  178930.250000   73179.250000  57.000000   
50%    1993.000000  181139.500000  193018.000000   89055.000000  70.500000   
75%    1999.500000  193353.250000  209295.500000  102369.500000  73.750000   
max    2008.000000  213314.000000  230872.000000  131304.000000  78.000000   

            felons        noncit     overseas  osvoters   VEPratio   VAPratio  
count    14.000000     14.000000    14.000000       1.0  14.000000  14.000000  
mean   2176.642857  12229.142857  2745.714286     263.0  48.949371  45.456576  
std     875.275131   4470.780002   840.053975       NaN   8.885781   8.125228  
min     802.000000   5756.000000  1803.000000     263.0  38.093162  34.831693  
25%    1423.750000   8591.500000  2236.000000     263.0  39.914544  36.566553  
50%    2312.000000  11972.000000  2458.500000     263.0  52.213204  48.437934  
75%    3042.250000  15910.500000  2937.000000     263.0  54.992570  52.413042  
max    3168.000000  19392.000000  4972.000000     263.0  61.554328  55.674090

In [9]:
diff = table['VEPratio'].mean() - table['VAPratio'].mean()
diff

3.492794466190091

### 投票率の平均

VEP：48.95％

VAP：45.45％

定義より、VAPのはVEPより大きいため、VEPで算出した投票率はVAPのものに比べて約3.49％高くなる

## 3. ANESに対しての投票率の差を求める
### ANES v.s. VEP
ANESの出した投票率とVEPをもとに算出した投票率の差の平均を求める

In [10]:
table['VEPvsANES'] = table['ANES'] -table['VEPratio']
table['VEPvsANES']

0     16.804491
1     17.862987
2     18.751404
3     14.858846
4     17.231520
5      8.581054
6     16.886160
7     14.873745
8     21.342072
9     13.906838
10    18.775507
11    22.489359
12    16.899156
13    16.445672
Name: VEPvsANES, dtype: float64

In [11]:
table['VEPvsANES'].describe()

count    14.000000
mean     16.836344
std       3.346069
min       8.581054
25%      15.266727
50%      16.892658
75%      18.529300
max      22.489359
Name: VEPvsANES, dtype: float64

約16.84％の平均差があり、約8.58%から約22.49%の範囲内で差があることが分かる。

### ANES v.s. VAP

ANESの出した投票率とVAPをもとに算出した投票率の差の平均を求める

In [12]:
table['VAPvsANES'] = table['ANES'] -table['VAPratio']
table['VAPvsANES']

0     18.960276
1     19.754777
2     21.462519
3     16.921549
4     20.277398
5     11.061159
6     20.959026
7     17.969142
8     25.466237
9     17.168307
10    23.657895
11    26.171504
12    22.452234
13    22.325910
Name: VAPvsANES, dtype: float64

In [13]:
table['VAPvsANES'].describe()

count    14.000000
mean     20.329138
std       3.893322
min      11.061159
25%      18.216926
50%      20.618212
75%      22.420653
max      26.171504
Name: VAPvsANES, dtype: float64

VAPの場合、約20.33％の平均差があり、約11.06%から約26.17%の範囲内で差があることが分かる

集計方法に偏りがないとするなら、VEPのほうがANESに近くまたVAPよりも差の範囲が狭い

## 4. 大統領選と中間選挙での比較
### 大統領選におけるVEPとANESの比較
大統領選は四年おきのため閏年のみを対象とする



In [14]:
elecPres = table[table['year'] % 4 == 0]
elecPres

year     VEP     VAP   total  ANES  felons  noncit  overseas  osvoters  \
0   1980  159635  164445   86515    71     802    5756      1803       NaN   
2   1984  167702  173995   92653    74    1165    7482      2361       NaN   
4   1988  173579  181955   91595    70    1594    9280      2257       NaN   
6   1992  179656  190778  104405    75    2183   11447      2418       NaN   
8   1996  186347  200016   96263    73    2586   13601      2499       NaN   
10  2000  194331  210623  105375    73    3083   16218      2937       NaN   
12  2004  203483  220336  122295    77    3158   18068      3862       NaN   
13  2008  213314  230872  131304    78    3145   19392      4972     263.0   

     VEPratio   VAPratio  VEPvsANES  VAPvsANES  
0   54.195509  52.039724  16.804491  18.960276  
2   55.248596  52.537481  18.751404  21.462519  
4   52.768480  49.722602  17.231520  20.277398  
6   58.113840  54.040974  16.886160  20.959026  
8   51.657928  47.533763  21.342072  25.466237  
10  54.224493  49.342105  18.775507  23.657895  
12  60.100844  54.547766  16.899156  22.452234  
13  61.554328  55.674090  16.445672  22.325910

In [15]:
elecPres.describe()

year            VEP            VAP         total       ANES  \
count     8.000000       8.000000       8.000000       8.00000   8.000000   
mean   1994.000000  184755.875000  196627.500000  103800.62500  73.875000   
std       9.797959   18257.329976   23113.313955   15721.58169   2.748376   
min    1980.000000  159635.000000  164445.000000   86515.00000  70.000000   
25%    1987.000000  172109.750000  179965.000000   92388.50000  72.500000   
50%    1994.000000  183001.500000  195397.000000  100334.00000  73.500000   
75%    2001.000000  196619.000000  213051.250000  109605.00000  75.500000   
max    2008.000000  213314.000000  230872.000000  131304.00000  78.000000   

            felons        noncit     overseas  osvoters   VEPratio   VAPratio  \
count     8.000000      8.000000     8.000000       1.0   8.000000   8.000000   
mean   2214.500000  12655.500000  2888.625000     263.0  55.983002  51.929813   
std     935.868275   5006.904376  1036.440742       NaN   3.553731   2.843662   
min     802.000000   5756.000000  1803.000000     263.0  51.657928  47.533763   
25%    1486.750000   8830.500000  2335.000000     263.0  53.838751  49.627478   
50%    2384.500000  12524.000000  2458.500000     263.0  54.736544  52.288602   
75%    3098.500000  16680.500000  3168.250000     263.0  58.610591  54.167672   
max    3158.000000  19392.000000  4972.000000     263.0  61.554328  55.674090   

       VEPvsANES  VAPvsANES  
count   8.000000   8.000000  
mean   17.891998  21.945187  
std     1.653152   2.022371  
min    16.445672  18.960276  
25%    16.865743  20.788619  
50%    17.065338  21.894215  
75%    18.757430  22.753649  
max    21.342072  25.466237

大統領選におけるVEP投票率の平均は約55.98%
またANESの平均は約73.88%

VEPとANESの差は平均して約17.89%ある

### 中間選挙におけるVEPとANESの比較
中間選挙では閏年以外の年の結果を対象とする

In [16]:
 midElec = table[table['year'] % 4 != 0]
 midElec

year     VEP     VAP  total  ANES  felons  noncit  overseas  osvoters  \
1   1982  160467  166028  67616    60     960    6641      1982       NaN   
3   1986  170396  177922  64991    53    1367    8362      2216       NaN   
5   1990  176629  186159  67859    47    1901   10239      2659       NaN   
7   1994  182623  195258  75106    56    2441   12497      2229       NaN   
9   1998  190420  205313  72537    52    2920   14988      2937       NaN   
11  2002  198382  215462  78382    62    3168   17237      3308       NaN   

     VEPratio   VAPratio  VEPvsANES  VAPvsANES  
1   42.137013  40.245223  17.862987  19.754777  
3   38.141154  36.078451  14.858846  16.921549  
5   38.418946  35.938841   8.581054  11.061159  
7   41.126255  38.030858  14.873745  17.969142  
9   38.093162  34.831693  13.906838  17.168307  
11  39.510641  35.828496  22.489359  26.171504

In [17]:
midElec.describe()

year            VEP            VAP         total      ANES  \
count     6.000000       6.000000       6.000000      6.000000   6.00000   
mean   1992.000000  179819.500000  191023.666667  71081.833333  55.00000   
std       7.483315   13698.886944   18107.726336   5120.561821   5.51362   
min    1982.000000  160467.000000  166028.000000  64991.000000  47.00000   
25%    1987.000000  171954.250000  179981.250000  67676.750000  52.25000   
50%    1992.000000  179626.000000  190708.500000  70198.000000  54.50000   
75%    1997.000000  188470.750000  202799.250000  74463.750000  59.00000   
max    2002.000000  198382.000000  215462.000000  78382.000000  62.00000   

            felons        noncit     overseas  osvoters   VEPratio   VAPratio  \
count     6.000000      6.000000     6.000000       0.0   6.000000   6.000000   
mean   2126.166667  11660.666667  2555.166667       NaN  39.571195  36.825594   
std     871.973260   4024.064744   504.633497       NaN   1.706609   1.972862   
min     960.000000   6641.000000  1982.000000       NaN  38.093162  34.831693   
25%    1500.500000   8831.250000  2219.250000       NaN  38.210602  35.856082   
50%    2171.000000  11368.000000  2444.000000       NaN  38.964794  36.008646   
75%    2800.250000  14365.250000  2867.500000       NaN  40.722351  37.542756   
max    3168.000000  17237.000000  3308.000000       NaN  42.137013  40.245223   

       VEPvsANES  VAPvsANES  
count   6.000000   6.000000  
mean   15.428805  18.174406  
std     4.596035   4.892309  
min     8.581054  11.061159  
25%    14.144840  16.983239  
50%    14.866296  17.568725  
75%    17.115677  19.308368  
max    22.489359  26.171504

中間選挙におけるVEP投票率の平均は約39.57%
またANESの平均は約55.00%

VEPとANESの差は平均して約15.43%ある

### 選挙の種類による投票率の比較

In [18]:
 diff = pd.DataFrame({
     'diff': [elecPres['ANES'].mean() - midElec['ANES'].mean(),
              elecPres['VEPratio'].mean() - midElec['VEPratio'].mean(),
              elecPres['VEPvsANES'].mean() - midElec['VEPvsANES'].mean()]},
     index = ['ANES','VEPratio','VEPvsANES']
 )
 
 diff

diff
ANES       18.875000
VEPratio   16.411807
VEPvsANES   2.463193

大統領選と中間選挙におけるANESの差の平均は約18.88%VEP

VEP投票率は約16.41%の差がある。

またVEP投票率とANES推定値に約2.46%の平均差がある。

この結果より大統領選では中間選挙に比べANES投票率は高くなる傾向が存在する。

## 5. 選挙の年を基準とし、前期後期で比較する。
### テーブルを2つに分割する

In [19]:
first = table[table['year'] < table['year'].mean()]
first

year     VEP     VAP   total  ANES  felons  noncit  overseas  osvoters  \
0  1980  159635  164445   86515    71     802    5756      1803       NaN   
1  1982  160467  166028   67616    60     960    6641      1982       NaN   
2  1984  167702  173995   92653    74    1165    7482      2361       NaN   
3  1986  170396  177922   64991    53    1367    8362      2216       NaN   
4  1988  173579  181955   91595    70    1594    9280      2257       NaN   
5  1990  176629  186159   67859    47    1901   10239      2659       NaN   
6  1992  179656  190778  104405    75    2183   11447      2418       NaN   

    VEPratio   VAPratio  VEPvsANES  VAPvsANES  
0  54.195509  52.039724  16.804491  18.960276  
1  42.137013  40.245223  17.862987  19.754777  
2  55.248596  52.537481  18.751404  21.462519  
3  38.141154  36.078451  14.858846  16.921549  
4  52.768480  49.722602  17.231520  20.277398  
5  38.418946  35.938841   8.581054  11.061159  
6  58.113840  54.040974  16.886160  20.959026

In [20]:
second = table[table['year'] >= table['year'].mean()]
second

year     VEP     VAP   total  ANES  felons  noncit  overseas  osvoters  \
7   1994  182623  195258   75106    56    2441   12497      2229       NaN   
8   1996  186347  200016   96263    73    2586   13601      2499       NaN   
9   1998  190420  205313   72537    52    2920   14988      2937       NaN   
10  2000  194331  210623  105375    73    3083   16218      2937       NaN   
11  2002  198382  215462   78382    62    3168   17237      3308       NaN   
12  2004  203483  220336  122295    77    3158   18068      3862       NaN   
13  2008  213314  230872  131304    78    3145   19392      4972     263.0   

     VEPratio   VAPratio  VEPvsANES  VAPvsANES  
7   41.126255  38.030858  14.873745  17.969142  
8   51.657928  47.533763  21.342072  25.466237  
9   38.093162  34.831693  13.906838  17.168307  
10  54.224493  49.342105  18.775507  23.657895  
11  39.510641  35.828496  22.489359  26.171504  
12  60.100844  54.547766  16.899156  22.452234  
13  61.554328  55.674090  16.445672  22.325910

### 上半期、下半期それぞれのVEPとANES投票率の平均差を求める

In [21]:
termDiff = pd.DataFrame({
     'first': [first['ANES'].mean(), 
              first['VEPratio'].mean(), 
              first['VEPvsANES'].mean()],
     'second': [second['ANES'].mean(),
                second['VEPratio'].mean(),
                second['VEPvsANES'].mean()]},
     index = ['ANES', 'VEPratio', 'VEPvsANES']
)

termDiff['diff'] = termDiff['second']-termDiff['first']
termDiff

first     second      diff
ANES       64.285714  67.285714  3.000000
VEPratio   48.431934  49.466807  1.034874
VEPvsANES  15.853781  17.818907  1.965126

上半期に比べ、ANESは3%の上昇

VEP投票率は約1.03%上昇

VEP投票率とANES投票率は約1.97%の上昇が認められた。

よって、ANES投票率のバイアスは時間が進むとともに増加する傾向が認められた。

## 6. 補正済みVAP投票率と比較
### 補正済みVAPを求める
VAPからfelonsと国籍不保持者を除き、補正済みVAP投票率を求める

In [22]:
table['corrVAP'] = table['VAP'] - (table['felons'] + table['noncit'])

In [23]:
temp = table['osvoters'].fillna(0).astype(int)
table['corrTotal'] = table['total'] - temp
table['corrVAPratio'] = table['corrTotal']/table['corrVAP']*100
table['corrVAPvsVAP'] = table['corrVAPratio'] - table['VAPratio']
table['corrVAPvsVEP'] = table['corrVAPratio'] - table['VEPratio']
table['corrVAPvsANES'] = table['ANES'] - table['corrVAPratio']

table.iloc[:,[0,-6,-5,-4,-3,-2,-1]]

year  corrVAP  corrTotal  corrVAPratio  corrVAPvsVAP  corrVAPvsVEP  \
0   1980   157887      86515     54.795518      2.755795      0.600010   
1   1982   158427      67616     42.679594      2.434370      0.542581   
2   1984   165348      92653     56.035150      3.497669      0.786554   
3   1986   168193      64991     38.640728      2.562277      0.499575   
4   1988   171081      91595     53.538967      3.816365      0.770487   
5   1990   174019      67859     38.995167      3.056327      0.576221   
6   1992   177148     104405     58.936595      4.895621      0.822756   
7   1994   180320      75106     41.651508      3.620651      0.525254   
8   1996   183829      96263     52.365514      4.831751      0.707585   
9   1998   187405      72537     38.706011      3.874318      0.612849   
10  2000   191322     105375     55.077304      5.735199      0.852811   
11  2002   195057      78382     40.184151      4.355656      0.673510   
12  2004   199110     122295     61.420823      6.873057      1.319979   
13  2008   208335     131041     62.899177      7.225087      1.344848   

    corrVAPvsANES  
0       16.204482  
1       17.320406  
2       17.964850  
3       14.359272  
4       16.461033  
5        8.004833  
6       16.063405  
7       14.348492  
8       20.634486  
9       13.293989  
10      17.922696  
11      21.815849  
12      15.579177  
13      15.100823

### 補正済みVAP投票率とその他の投票率の比較

In [24]:
comp = pd.DataFrame({
     'mean': [table['ANES'].mean(),
              table['VEPratio'].mean(),
              table['VAPratio'].mean(),
              table['corrVAPratio'].mean(),   
              table['VEPratio'].mean() - table['VAPratio'].mean(),
              table['VAPvsANES'].mean(),
              table['corrVAPvsVAP'].mean(),
              table['corrVAPvsVEP'].mean(),
              table['corrVAPvsANES'].mean()]},
     index = ['ANES','VEPratio','VAPratio','corrVAPratio', 
              'VAPvsVEP', 'VAPvsANES',
              'corrVAPvsVAP','corrVAPvsVEP','corrVAPvsANES']
 )

comp

mean
ANES           65.785714
VEPratio       48.949371
VAPratio       45.456576
corrVAPratio   49.709015
VAPvsVEP        3.492794
VAPvsANES      20.329138
corrVAPvsVAP    4.252439
corrVAPvsVEP    0.759644
corrVAPvsANES  16.076699

補正済みVAPでは投票権を持たない人口をカウントしていないため、VAPに比べて投票率が約4.25%上昇している。

またVAPとVEPの補正前平均差が約3.49%あったのに比べ、補正後には約0.76%まで差が縮まっている。

同じくAMES投票率との平均差も補正前の約20.33%から補正後は約16.08%の差へと縮まっている。

これらの結果から、VAPを補正することで補正前の投票率に比べより精度の高い投票率を算出できると考えられる。